File: core/src/main/java/com/tuuzed/androidx/recyclerview/adapter/internal/ItemTypePool.java
package com.tuuzed.androidx.recyclerview.adapter.internal;

import com.tuuzed.androidx.recyclerview.adapter.ItemTypeViewBinder;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;

public final class ItemTypePool {
    @NonNull
    private List<ItemTypeViewBinder<?, ?>> mItemTypeViewBinders;
    @NonNull
    private List<Class<?>> mItemTypes;

    public ItemTypePool(int initCapacity) {
        mItemTypeViewBinders = new ArrayList<>(initCapacity);
        mItemTypes = new ArrayList<>(initCapacity);
    }

    public <T> void register(Class<T> itemType, ItemTypeViewBinder<T, ?> itemTypeBinder) {
        int index = indexOfItemType(itemType);
        if (index == -1) {
            mItemTypes.add(itemType);
            mItemTypeViewBinders.add(itemTypeBinder);
        } else {
            mItemTypeViewBinders.set(index, itemTypeBinder);
        }
    }

    public int indexOf(@NonNull Class<?> itemType) {
        int itemViewType = indexOfItemType(itemType);
        if (itemViewType == -1) {
            throw new IllegalStateException("Not registered (" + itemType.getName() + ") type.");
        }
        return itemViewType;
    }

    @NonNull
    public ItemTypeViewBinder<?, ?> findItemTypeViewBinder(int viewType) {
        if (mItemTypeViewBinders.size() < viewType - 1) {
            throw new IllegalStateException("Not viewType (" + viewType + ") type.");
        }
        return mItemTypeViewBinders.get(viewType);
    }

    private int indexOfItemType(@NonNull Class<?> itemType) {
        return mItemTypes.indexOf(itemType);
    }

}

File: adapter/src/main/java/com/tuuzed/recyclerview/adapter/ItemTypePool.kt
package com.tuuzed.recyclerview.adapter

import androidx.recyclerview.widget.RecyclerView
import java.util.*

internal class ItemTypePool(initCapacity: Int) {
    private val itemTypeViewBinders: MutableList<ItemTypeViewBinder<*, *>>
    private val itemTypes: MutableList<Class<*>>

    init {
        itemTypeViewBinders = ArrayList(initCapacity)
        itemTypes = ArrayList(initCapacity)
    }

    fun <T> register(itemType: Class<in T>, itemTypeBinder: ItemTypeViewBinder<T, *>) {
        val index = indexOfItemType(itemType)
        if (index == -1) {
            itemTypes.add(itemType)
            itemTypeViewBinders.add(itemTypeBinder)
        } else {
            itemTypeViewBinders[index] = itemTypeBinder
        }
    }

    fun indexOf(itemType: Class<*>): Int {
        val itemViewType = indexOfItemType(itemType)
        if (itemViewType == -1) {
            throw IllegalStateException("Not registered (" + itemType.name + ") type.")
        }
        return itemViewType
    }

    fun findItemTypeViewBinder(viewType: Int): ItemTypeViewBinder<Any, RecyclerView.ViewHolder> {
        if (itemTypeViewBinders.size < viewType - 1) {
            throw IllegalStateException("Not viewType ($viewType) type.")
        }
        @Suppress("UNCHECKED_CAST")
        return itemTypeViewBinders[viewType] as ItemTypeViewBinder<Any, RecyclerView.ViewHolder>
    }

    private fun indexOfItemType(itemType: Class<*>): Int {
        return itemTypes.indexOf(itemType)
    }


}